File: plugins/collector/event/apps/database/mysql.go
package database

import (
	"collector/cache/process"
	"collector/event/apps"
	"regexp"
	"strings"
)

type Mysql struct {
	version string
	r       *regexp.Regexp
}

func (Mysql) Name() string { return "mysql" }

func (Mysql) Type() string { return "database" }

func (m Mysql) Version() string { return m.version }

func (Mysql) Match(p *process.Process) bool { return p.Name == "mysqld" }

func (m *Mysql) Run(p *process.Process) (mapping map[string]string, err error) {
	if m.r == nil {
		m.r = regexp.MustCompile(`Ver\s(\d+\.)+\d+`)
	}
	result, err := apps.Execute(p, "-V")
	if err != nil {
		return nil, err
	}
	str := m.r.FindString(result)
	if str == "" {
		err = apps.ErrVersionNotFound
		return
	}
	m.version = strings.TrimPrefix(str, "Ver ")
	return
}

func init() {
	apps.Regist(&Mysql{})
}

File: plugins/collector/src/event/apps/database/mysql.rs
use crate::event::apps::{execute, find_version, AppProc, IApp};
use std::collections::HashMap;

pub struct Mysql {
    version: String,
}

impl Mysql {
    pub fn new() -> Self { Mysql { version: String::new() } }
}

impl IApp for Mysql {
    fn name(&self)     -> &'static str { "mysql" }
    fn app_type(&self) -> &'static str { "database" }
    fn version(&self)  -> &str         { &self.version }

    fn matches(&mut self, p: &AppProc) -> bool { p.name == "mysqld" }

    fn run(&mut self, p: &AppProc) -> anyhow::Result<HashMap<String, String>> {
        let out = execute(p, &["-V"])?;
        // "mysqld  Ver 8.0.30 Distrib 8.0.30 …"
        if let Some(ver_part) = out.split("Ver ").nth(1) {
            self.version = find_version(ver_part);
        }
        if self.version.is_empty() { self.version = find_version(&out); }
        if self.version.is_empty() {
            return Err(anyhow::anyhow!("version not found"));
        }
        Ok(HashMap::new())
    }
}
